File: problems/49-group-anagrams/solution.go
package main

import (
	"sort"
)

func groupAnagrams(strs []string) [][]string {
	group := make(map[string][]string)

	for _, v := range strs {
		letters := []rune(v)

		sort.Slice(letters, func(i, j int) bool {
			return letters[i] < letters[j]
		})

		word := string(letters)

		group[word] = append(group[word], v)
	}

	res := make([][]string, 0, len(group))

	for _, v := range group {
		res = append(res, v)
	}

	return res
}

File: problems/49-group-anagrams/solution.ts
function groupAnagrams(strs: string[]): string[][] {
  const map = new Map<string, string[]>();

  for (let i = 0; i < strs.length; i++) {
    let sortedLetters = strs[i].split("").sort().join("");

    map.has(sortedLetters)
      ? map.get(sortedLetters)!.push(strs[i])
      : map.set(sortedLetters, [strs[i]]);
  }

  return [...map.values()]
}

export { groupAnagrams };
